### PR TITLE
USERPROFILE is not always defined, use psutils instead

### PR DESCRIPTION
**Pull request by mattip**
## Why are these changes needed?
Occasionally, tests are crashing. Looking through the logs, I see that the reporter is crashing because `USERPROFILE` is not defined, the check fails, and the process crashes. I am not sure how this happens, but we can use `psutils.disk_partitions` which does not fail.

This might be the source of some of the test flakiness.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by mattip** [comment]
I think the undefined USERPROFILE is coming from the call to [Process()](https://github.com/ray-project/ray/blob/5d3fc5a6770580f4c218b945804f66b0eb77c374/src/ray/raylet/worker_pool.cc#L470) with a [very limited env](https://github.com/ray-project/ray/blob/5d3fc5a6770580f4c218b945804f66b0eb77c374/src/ray/raylet/worker_pool.cc#L396).

**Comment by mattip** [comment]
First add USERPROFILE to bazel, that might solve the problem?

File: src/ray/raylet/worker_pool.cc [hunk 1]
```diff
@@ -402,6 +402,9 @@ std::tuple<Process, StartupToken> WorkerPool::StartWorkerProcess(
 #elif defined(__linux__)
       static const std::string kLibraryPathEnvName = "LD_LIBRARY_PATH";
 #elif defined(_WIN32)
+      // USERPROFILE is used in pathlib.Path.home()
```
Comment (pcmoritz) [inline src/ray/raylet/worker_pool.cc:405]: Does this actually make a difference / what happens if we don't add this? This codepath is only taken for C++ I think, so it shouldn't make a difference for Python :)

File: src/ray/util/process.cc [hunk 2]
```diff
@@ -142,9 +142,16 @@ class ProcessFD {
         (void)cmd.c_str();  // We'll need this to be null-terminated (but mutable) below
         TCHAR *cmdline = &*cmd.begin();
         STARTUPINFO si = {sizeof(si)};
-        RAY_UNUSED(
-            new_env_block.c_str());  // Ensure there's a final terminator for Windows
-        char *const envp = &new_env_block[0];
+        LPVOID envp = NULL;
+        if (!new_env_block.empty()) {
+          if (new_env.find("PATH") == new_env.end()) {
```
Comment (pcmoritz) [inline src/ray/util/process.cc:147]: Why was this needed?

File: src/ray/raylet/worker_pool.cc (hunk 1 above)
Comment (mattip) [reply src/ray/raylet/worker_pool.cc:405]: > Does this actually make a difference

Apparently not, since tests are still failing. I am grasping at straws trying to figure out why. One theory I had was that the environment is not being set up properly for windows. It seems, grepping in the virtualenv I use to run tests, that `USERPROFILE` is expected to exist when running python programs, which inherit the environment from the enclosing worker. If `USERPROFILE` is not set in the environment, it will not be inherited, and things like `pathlib.Path.home()` will fail, which I did see in logs locally.

File: src/ray/util/process.cc (hunk 2 above)
Comment (mattip) [reply src/ray/util/process.cc:147]: There are two changes here:
- If `new_env_block` is empty, setting `env_p = ""` and calling `CreateProcess(..., env_p)` is equivalent to calling `Popen(env={})` which will cause the process to be started without a `PATH` and without `USERPROFILE`. So check for `new_env_block` before setting `env_p` to not be NULL (which will then copy the enclosing process's environment).
- Check that the `new_env` at least has `PATH` set, to try to preemptively crash

**Comment by mattip** [comment]
Closing. This was done in PR #22449.
